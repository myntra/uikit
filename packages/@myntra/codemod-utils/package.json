{
  "name": "@myntra/codemod-utils",
  "version": "0.1.0-alpha.2",
  "description": "High level helper function for AST transformations.",
  "main": "src/index.js",
  "author": "Rahul Kadyan <hi@znck.me>",
  "license": "UNLICENSED",
  "repository": "https://bitbucket.org/myntra/uikit/src/master/packages/@myntra/codemod-utils",
  "publishConfig": {
    "registry": "http://registry.myntra.com:8000"
  },
  "files": [
    "src/"
  ],
  "dependencies": {
    "debug": "^3.1.0",
    "eslint": "^4.19.1",
    "execa": "^0.10.0",
    "glob": "^7.1.2",
    "jscodeshift": "^0.5.1",
    "jsdoc-api": "^4.0.3",
    "prettier": "^1.13.4"
  }
}
