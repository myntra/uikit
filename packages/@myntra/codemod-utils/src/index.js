import debug from 'debug'
import prettier from 'prettier'

const d = debug('@myntra/codemod-utils')

export function createHelper(file, api) {
  const j = api.jscodeshift
  const root = j(file.source)
  const h = helpers(j, root)

  return { j, h, root }
}

export default function helpers(j, root) {
  function first(nodes) {
    if (!nodes || nodes.size() === 0) return

    d('first: ' + 0 + ' of ' + nodes.size())

    return nodes.paths()[0]
  }

  function findLastNonRelativeImportStatement() {
    const result = root.find(
      j.Statement,
      node => node.type === 'ImportDeclaration' && !node.source.value.startsWith('.')
    )

    if (result.size()) return result.paths()[result.size() - 1]
  }

  function findFirstNonImportStatement() {
    const result = root.find(j.Statement, node => node.type !== 'ImportDeclaration')

    if (result.size()) return first(result)
  }

  function findImport(source) {
    return root.find(j.ImportDeclaration).filter(decl => decl.value.source.value === source)
  }

  function insertAtEnd(node) {
    root
      .find(j.Program)
      .get()
      .value.body.push(node)
  }

  function insertAfterImports(node) {
    const pos = findFirstNonImportStatement()

    if (pos) {
      j(pos).insertBefore(node)
    } else {
      insertAtEnd(node)
    }
  }

  function insertAfterNonRelativeImports(node) {
    const pos = findLastNonRelativeImportStatement()

    if (pos) {
      j(pos).insertAfter(node)
    } else {
      insertAtEnd(node)
    }
  }

  function addImport(source, name, local, named = false) {
    const existing = first(findImport(source))

    const globalCollisions = root
      .find(j.Identifier, n => n.name === local && n.type !== (named ? 'ImportSpecifier' : 'ImportDefaultSpecifier'))
      .paths()
      .filter(
        n =>
          j(n)
            .closestScope()
            .get().value.type === 'Program'
      )

    if (globalCollisions.length) {
      throw Error(local + ' is already defined in scope.')
    }

    if (!existing) {
      insertAfterNonRelativeImports(
        j.importDeclaration(
          [
            named
              ? j.importSpecifier(j.identifier(name), j.identifier(local))
              : j.importDefaultSpecifier(j.identifier(local))
          ],
          j.literal(source)
        )
      )

      return // inserted. return early.
    }

    const specifier = named
      ? existing.value.specifiers.find(s => s.type === 'ImportSpecifier' && s.imported.name === local)
      : existing.value.specifiers.find(s => s.type === 'ImportDefaultSpecifier')

    if (!specifier) {
      // import <default>, { name as local, ... } from source
      existing.value.specifiers.push(
        named
          ? j.importSpecifier(j.identifier(name), j.identifier(local || name))
          : j.importDefaultSpecifier(j.identifier(local || name))
      )
    } else if (specifier.local.name !== local) {
      insertAfterImports(
        j.variableDeclaration('const', [j.variableDeclarator(j.identifier(name), j.identifier(specifier.local.name))])
      )
    }
  }

  /**
   * Add named import.
   * e.g. import { name as local } from 'source'
   *
   * @param {string} source
   * @param {string} name
   * @param {string} [local]
   * @returns {void}
   */
  function addNamedImport(source, name, local) {
    addImport(source, name, local || name, true)
  }

  /**
   * Add default import.
   * e.g. import local from 'source'
   *
   * @param {string} source
   * @param {string} local
   * @return {void}
   */
  function addDefaultImport(source, local) {
    addImport(source, undefined, local, false)
  }

  function getPropInteropNode(localComponentName) {
    const name = 'interopPropTransformer' + localComponentName + '$0'
    const result = root.find(j.VariableDeclaration, node => node.declarations[0].id.name === name)

    if (result.size()) return first(result).value

    addNamedImport('@myntra/uikit', 'interopPropTransformer')

    const interop = j.variableDeclaration('const', [
      j.variableDeclarator(
        j.identifier(name),
        j.callExpression(j.identifier('interopPropTransformer'), [j.objectExpression([]), j.objectExpression([])])
      )
    ])

    insertAfterImports(interop)

    return interop
  }

  function addPropInteropMapping(node, from, to) {
    const decl = node.declarations[0]
    const mapping = decl.init.arguments[0]

    const existing = mapping.properties.find(it => it.key.name === from)

    if (!existing) {
      mapping.properties.push(j.objectProperty(j.identifier(from), j.literal(to)))
    }
  }

  function addPropInteropCoercion(node, from, fn) {
    const decl = node.declarations[0]
    const coercions = decl.init.arguments[1]

    const existing = coercions.properties.find(it => it.key.name === from)

    if (!existing) {
      coercions.properties.push(
        j.objectProperty(j.identifier(from), first(j('const foo = ' + fn.toString()).find(j.FunctionExpression)).value)
      )
    }
  }

  function renameProp(localComponentName, oldPropName, newPropName) {
    root
      .find(j.JSXOpeningElement, { name: { type: 'JSXIdentifier', name: localComponentName } })
      .replaceWith(nodePath => {
        nodePath.node.attributes.forEach((attribute, index) => {
          if (attribute.type === 'JSXAttribute' && attribute.name.name === oldPropName) {
            nodePath.node.attributes.splice(index, 1, j.jsxAttribute(j.jsxIdentifier(newPropName), attribute.value))
          } else if (attribute.type === 'JSXSpreadAttribute') {
            const interop = getPropInteropNode(localComponentName)
            const name = interop.declarations[0].id.name

            if (!(attribute.argument.type === 'CallExpression' && attribute.argument.callee.name === name)) {
              attribute.argument = j.callExpression(j.identifier(name), [attribute.argument])
            }

            addPropInteropMapping(interop, oldPropName, newPropName)
          }
        })

        return nodePath.node
      })
  }

  function coerceProp(localComponentName, prop, fn) {
    root
      .find(j.JSXOpeningElement, { name: { type: 'JSXIdentifier', name: localComponentName } })
      .replaceWith(nodePath => {
        nodePath.node.attributes.forEach((attribute, index) => {
          if (
            attribute.type === 'JSXSpreadAttribute' ||
            (attribute.type === 'JSXAttribute' && attribute.name.name === prop)
          ) {
            const interop = getPropInteropNode(localComponentName)
            const name = interop.declarations[0].id.name
            addPropInteropCoercion(interop, prop, fn)

            if (attribute.type === 'JSXAttribute') {
              attribute.value = j.jsxExpressionContainer(
                j.callExpression(
                  j.memberExpression(
                    j.memberExpression(j.identifier(name), j.identifier('coercions')),
                    j.identifier(prop)
                  ),
                  [attribute.value.type === 'JSXExpressionContainer' ? attribute.value.expression : attribute.value]
                )
              )
            } else if (!(attribute.argument.type === 'CallExpression' && attribute.argument.callee.name === name)) {
              attribute.argument = j.callExpression(j.identifier(name), [attribute.argument])
            }

            addPropInteropCoercion(interop, prop, fn)
          }
        })

        return nodePath.node
      })
  }

  function renameProps(localComponentName, propNames) {
    Object.entries(propNames).forEach(([from, to]) => renameProp(localComponentName, from, to))
  }

  function getDefaultImportLocalName(node) {
    const defaultSpecifier = node.value.specifiers.find(specifier => specifier.type === 'ImportDefaultSpecifier')

    if (defaultSpecifier) return defaultSpecifier.local.name

    throw new Error('No default import found')
  }

  return {
    first,
    findImport,
    findLastNonRelativeImportStatement,
    findFirstNonImportStatement,
    addNamedImport,
    addDefaultImport,
    renameProp,
    renameProps,
    coerceProp,
    getDefaultImportLocalName,
    toSource: () =>
      prettier.format(root.toSource({ quote: 'single', wrapColumn: 120, tabWidth: 2 }), {
        semi: false,
        singleQuote: true,
        parser: 'babylon'
      })
  }
}

/**
 *
 * @export
 * @param {string} dir
 * @param {string[]} filename
 * @param {any} [options={}]
 */
export function testCodeMod(dir, filename, options = {}) {
  const path = require('path')
  const glob = require('glob')
  const fs = require('fs')
  const file = path.resolve(dir, filename)
  const inputDir = path.resolve(dir, '__codemod__/input')
  const outputDir = path.resolve(dir, '__codemod__/output')
  const codemods = require(file)

  describe(filename, () => {
    function read(f) {
      return fs.readFileSync(f, 'utf8').toString()
    }
    for (const name in codemods) {
      describe(name, () => {
        const fixtures = glob.sync(name + '*.js', { cwd: inputDir })

        fixtures.forEach(fixture => {
          it(fixture, () => {
            const isNegative = fixture.includes('.fail.')
            const inputPath = path.resolve(inputDir, fixture)
            const outputPath = path.resolve(outputDir, fixture)

            const transform = codemods[name]
            let jscodeshift = require('jscodeshift/src/core')
            if (transform.parser) {
              jscodeshift = jscodeshift.withParser(transform.parser)
            }

            try {
              const output = transform(
                {
                  path: inputPath,
                  source: read(inputPath)
                },
                { jscodeshift, stats: () => {} },
                options || {}
              )
              if (isNegative) {
                expect(output).toBeFalsy()
              } else {
                expect(output.trim()).toEqual(read(outputPath).trim())
              }
            } catch (e) {
              if (!isNegative) throw e
            }
          })
        })
      })
    }
  })
}
