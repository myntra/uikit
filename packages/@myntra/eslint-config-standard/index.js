module.exports = {
  parser: 'babel-eslint',
  extends: [
    'standard',
    'plugin:react/recommended',
    'prettier',
    'prettier/react',
    'prettier/standard',
    'plugin:jest/recommended'
  ],
  plugins: ['react', 'prettier', 'standard', 'json', 'babel', 'jest', 'node'],
  parserOptions: {
    sourceType: 'module',
    ecmaFeatures: {
      jsx: true
    }
  },
  env: {
    es6: true,
    node: true,
    'jest/globals': true
  },
  rules: {
    'node/no-unpublished-bin': 2,
    'node/no-extraneous-import': 2,
    'node/no-unpublished-import': 2,
    'node/no-extraneous-require': 2,
    'node/no-unpublished-require': 2,
    'prettier/prettier': [
      'error',
      {
        bracketSpacing: true,
        jsxBracketSameLine: false,
        printWidth: 120,
        semi: false,

        /* `singleQuote: true`
         * IN PRETTIER IS SIMILAR TO THE ESLINT RULE:
         * `quotes: [error, single, { avoidEscape: true }],`
         */
        singleQuote: true,
        tabWidth: 2,
        trailingComma: 'none',
        useTabs: false,
        parser: 'babylon'
      }
    ]
  }
}
