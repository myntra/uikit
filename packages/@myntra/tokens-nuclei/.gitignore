*
!package.json
!tokens.yml
!.gitignore
