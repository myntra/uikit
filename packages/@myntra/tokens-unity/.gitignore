*
!package.json
!tokens.yml
!.gitignore
!reset.css
