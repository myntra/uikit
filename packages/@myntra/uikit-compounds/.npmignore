*.spec.js
__codemod__
__snapshots__
fixtures
