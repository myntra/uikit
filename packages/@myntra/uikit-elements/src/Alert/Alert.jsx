import React from 'react'
import PropTypes from 'prop-types'
import classnames from './Alert.module.css'
import { Icon } from '../index.js'

const ICONS = {
  error: 'exclamation-triangle',
  warning: 'exclamation-triangle',
  success: 'check-circle',
  info: 'info-circle'
}

/**
 The Alert component.
 @since 0.3.0
 @status EXPERIMENTAL
 @example
<Alert onClose={() => null}>An alert message.</Alert>
 */
function Alert({ className, type, noFill, onClose, children }) {
  return (
    <div className={classnames('alert', { 'no-fill': noFill }, type, className)}>
      <Icon className={classnames('icon', 'legend')} name={ICONS[type]} />
      <div className={classnames('content')}>{children}</div>
      {onClose && (
        <div className={classnames('close')} onClick={onClose} role="button" data-test-id="close">
          <Icon name="times" className={classnames('icon')} />
        </div>
      )}
    </div>
  )
}

Alert.propTypes = {
  /** @private */
  className: PropTypes.string,
  /**
   * Handler for change event
   * @function
   */
  onClose: PropTypes.func,
  /** Only border based Alert */
  noFill: PropTypes.bool,
  /** Type of Alert  */
  type: PropTypes.oneOf(['success', 'error', 'warning', 'info']),
  /** Children */
  children: PropTypes.any.isRequired
}

Alert.defaultProps = {
  noFill: false,
  type: 'error'
}

export default Alert
