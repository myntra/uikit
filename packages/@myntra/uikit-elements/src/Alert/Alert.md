```jsx render editor
function onClose() {}

<>
  <Alert type="information" onClose={onClose}>
    New recipes has been generated for this drop.
  </Alert>
  <Alert type="success" onClose={onClose}>
    Success! New manual recipe is created.
  </Alert>
  <Alert type="caution" onClose={onClose}>
    CAUTION! The image you are trying to upload has invalid format. It might not render well. <br />
    <Button type="link.inherit">Proceed Anyway</Button>
  </Alert>
  <Alert type="error" onClose={onClose}>
    ERROR! The drop you were trying to create has invalid inputs.
  </Alert>
</>
```

```jsx render editor
function onClose() {}

<>
  <Alert type="information" onClose={onClose} noFill>
    New recipes has been generated for this drop.
  </Alert>
  <Alert type="success" onClose={onClose} noFill>
    Success! New manual recipe is created.
  </Alert>
  <Alert type="caution" onClose={onClose} noFill>
    CAUTION! The image you are trying to upload has invalid format. It might not render well. <br />
    <Button type="link.inherit">Proceed Anyway</Button>
  </Alert>
  <Alert type="error" onClose={onClose} noFill>
    ERROR! The drop you were trying to create has invalid inputs.
  </Alert>
</>
```
