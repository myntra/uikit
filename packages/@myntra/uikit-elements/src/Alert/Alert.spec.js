import React from 'react'
import { shallow } from 'enzyme'

import { testCodeMod } from '@myntra/codemod-utils'

import Alert from './Alert'

testCodeMod(__dirname, 'Alert.codemod.js')

describe('Alert', () => {
  it('renders', () => {
    expect(shallow(<Alert>Alert</Alert>)).toMatchSnapshot()
  })
  it('triggers close event on cross click', () => {
    const handleClose = jest.fn()
    const wrapper = shallow(<Alert onClose={handleClose}>Alert</Alert>)
    wrapper.find('.close').simulate('click')
    expect(handleClose).toHaveBeenCalled()
  })
  it('renders unfilled alert', () => {
    const wrapper = shallow(<Alert noFill>Alert</Alert>)
    expect(wrapper.hasClass('no-fill')).toBe(true)
  })
})
