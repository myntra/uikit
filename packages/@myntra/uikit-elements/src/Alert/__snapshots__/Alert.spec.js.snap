// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert renders 1`] = `
<div
  className="alert error"
>
  <Icon
    className="icon legend"
    name="caution"
  />
  <div
    className="content"
  >
    Alert
  </div>
</div>
`;
